### ISSUE TITLE
Allocation of captured const variables is unstable.

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#26453]
The program below crashes with segfault because we when we first compile `foo` and `bar` we allocate all four variables (`a`, `b`, `c` and `d`) to the context.

When we compile `foo` the second time (with optimizations) we allocate only `c` and `d` to the context. This likely happens because parser folds away "${a}" and "${b}" as constant expressions when parsing `bar` on its own. 

This causes mismatch between a context that `bar` expects and that optimized version of `foo` produces.

```dart
foo() {
  const a = 1;
  const b = 2;

  var c = 3;
  var d = 4;

  bar() {
    if ("${a}" != "1") throw "failed";
    if ("${b}" != "2") throw "failed";
    if ("${c}" != "3") throw "failed";
    if ("${d}" != "4") throw "failed";
  }

  bar();
}

main() {
  for (var i = 0; i < 50000; i++) foo();
}
```

**Comment by mraleph** [comment]
@iposva-google @mhausner

**Comment by mhausner** [comment]
I can reproduce this, with and without background compilation. When optimization is turned off, the program completes without crash.

**Comment by mraleph** [comment]
@mhausner background optimization has nothing to do with this, it's purely parser-constant folding and scope analysis interaction

**Comment by mhausner** [comment]
Here is a slightly simplified version:

```
foo() {
  const a = 1;
  var c = 3;

  bar() {
    if ("${a}" != "1") throw "failed";
    if (c != 3) throw "failed";
  }

  bar();
}

main() {
  for (var i = 0; i < 500; i++) foo();
}
```

This crashes:
```
$ dart --optimization-counter-threshold=10  ~/tmp/y.dart 
Segmentation fault: 11
```

Interestingly, the crash does not happen when accessing const variable a directly instead of through the string interpolation. Replacing the first line in bar() with this:

```
if (a != 1) throw "failed";
```

Makes the crash disappear.

**Comment by mhausner** [comment]
@mraleph is right. Forcing a second compilation with the non-optimizing compiler produces the same crash:

```
$ dart --compile_all --optimization-counter-threshold=-1  ~/tmp/y.dart 
Unhandled exception:
failed
#0      foo.bar (file:///Users/hausner/tmp/y.dart:7:17)
...
```

**Comment by mhausner** [comment]
Working on a fix.

**Comment by mhausner** [comment]
Fix is out for review: https://codereview.chromium.org/1980193002/

**Comment by fsc8000** [comment]
This CL causes a crash / assertion failure in debug mode:

```
runtime/vm/parser.cc:7700: error: expected: v != NULL
Aborted (core dumped)
```

https://build.chromium.org/p/client.dart/builders/vm-linux-release-x64-optcounter-threshold-be/builds/823/steps/vm%20tests/logs/stdio

**Comment by mhausner** [comment]
Taking a look at crash. Didn't observe this on my machine.

**Comment by fsc8000** [comment]
Here is the command line I used in debug mode:

```
out/DebugX64/dart --optimization-counter-threshold=5 --ignore-unrecognized-flags --package-root=out/DebugX64/packages/ /media/SSD/dartgit/sdk/tests/corelib/map_test.dart
```

**Comment by mhausner** [comment]
This is fixed for good with 7daa2a8915af72ba9394fada91cb40402bff2624 .
